optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,643 @@
+from __future__ import annotations
+
+from collections import defaultdict
+import hashlib
+import itertools
+import math
+from typing import Any
+from typing import Callable
+from typing import Sequence
+import warnings
+
+import numpy as np
+
+import optuna
+from optuna.distributions import BaseDistribution
+from optuna.exceptions import ExperimentalWarning
+from optuna.samplers._base import _CONSTRAINTS_KEY
+from optuna.samplers._base import _process_constraints_after_trial
+from optuna.samplers._base import BaseSampler
+from optuna.samplers._random import RandomSampler
+from optuna.samplers._search_space import IntersectionSearchSpace
+from optuna.samplers.nsgaii._crossover import perform_crossover
+from optuna.samplers.nsgaii._crossovers._base import BaseCrossover
+from optuna.samplers.nsgaii._crossovers._uniform import UniformCrossover
+from optuna.samplers.nsgaii._sampler import _constrained_dominates
+from optuna.study import Study
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "nsga3:generation"
+_POPULATION_CACHE_KEY_PREFIX = "nsga3:population"
+# Define a coefficient for scaling intervals, used in _filter_inf() to replace +-inf
+_COEF = 3
```

Comment:
[Notes] The variable name of `_COEF` might be too general, but the code comment explains the usage clearly. So, it's not mandatory, but we might rename it in a follow-up PR.